### ISSUE TITLE
How to handle tables being wider than the page width?

### ISSUE THREAD

**Issue by Bazze** [bpampuch/pdfmake#441]
Hello!

I haven't been able to find any documentation or issues about how to handle tables with many columns, causing the table to be wider than the page width. How would one handle such a case? I know there is built-in support for automatically breaking the page if the table is getting close to the end of the page but is there similar features for breaking when a table gets too wide?

I'm currently in a situation where I have tables with a lot of columns causing it to get too wide. Would be nice if it just breaks and continues below just as for the page break feature.

Thanks.

**Comment by AaronBuxbaum** [comment]
I have the same bug; I think a good enhancement would be to add a flag that will throw an error if it extends off that page. That way, the consumer can handle the case on their end.

Breaking page horizontally would be a good feature as well.

**Comment by Bazze** [comment]
Yes, it needs to be handled one way or the other. I actually did an implementation that works for *my specific use-case*. When a table gets to wide it puts exceeding columns into a new table just below where the `headerRows` and `headerColumns` (new setting introduced by me) are repeated. I created a PR in my fork, https://github.com/Bazze/pdfmake/pull/1, where maybe you can find something of use.

**My fix for this is a school-book example of quick and dirty so you should NOT use it. I just had to get something to work quick, so I copy and pasted and scribbled some code down. I'm not proud of it.**

**Comment by DamienCassou** [comment]
Any progress on this old issue? We have the same problem.

**Comment by 9vimu9** [comment]
change page size

               `extend: 'pdfHtml5',
                orientation: 'landscape',//landscape give you more space
                pageSize: 'A0',//A0 is the largest A5 smallest(A0,A1,A2,A3,legal,A4,A5,letter))
                title:title`
datatatables link
[https://www.google.lk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=3&cad=rja&uact=8&ved=0ahUKEwj947HL4aPWAhXKpI8KHQ1yDNQQFgg2MAI&url=https%3A%2F%2Fdatatables.net%2Freference%2Fbutton%2Fpdf&usg=AFQjCNHOa2iXfZ2F_ERjCuA2kHjE5OW2zA](url)

more about standard paper sizes
[http://www.papersizes.org/images/a-series-paper-sizes-1.jpg](url)

**Comment by akhiljain1611** [comment]
any progress yet ?

**Comment by jasonvasquez** [comment]
We have the same issue.  Is there anyway to programatically determine that the table has been truncated, or to calculate it's final width to see if it's too wide?  I'm fine for now if we could know it was going to happen, and then reshape the content accordingly.

Thanks!

**Comment by ValeSauer** [comment]
```javascript
    calculateTableFontSize = function (colCount) {
        var defaultSize = docDefinition.defaultStyle.fontSize; // default = 10
        var size = defaultSize
        if (defaultSize) {
            size = 18 - colCount
            if (size > defaultSize)
                size = defaultSize
        }
        return size
    }

    if (table.table.body[0] && table.table.body[0].length) {
        docDefinition.styles.tableCell.fontSize = calculateTableFontSize(table.table.body[0].length) //add style: 'tableCell' to every table cell
    }
    docDefinition.content.push(table)
```
As my tables are usually only slightly too wide (max. 200% of page width) I can scale fontSize in order to make tables with too many columns fit to the page. Perhaps this workaround helps anyone.  However this does only take into account number of columns, not the actual width of the column.

**Comment by ShubhamOjha** [comment]
any work  around?

**Comment by srallapalli** [comment]
any updates on this issue?

**Comment by GiordanoArman** [comment]
Much needed by me too!

**Comment by axelisr** [comment]
Any updates on this? At least any way of knowing when table will overflow?

**Comment by nmenke** [comment]
You could use DocMeasure.measureTable to get the actual size and compare it to the page width and height.

**Comment by ajperez81** [comment]
> You could use DocMeasure.measureTable to get the actual size and compare it to the page width and height.

That seems to be the most reasonable way to workaround this issue. Is there any way to access `DocMeasure.measureTable` from outside pdfmake? Or would I need to modify pdfmake itself? I don't see any way directly from the `pdfmake` class, but maybe there are other objects exported aswell?

**Comment by nmenke** [comment]
If you don't use the bundled version, you can access everything needed. Just bundle it yourself.

**Comment by MussaCharles** [comment]
@Bazze  Did you find any workaround on this issue?

**Comment by Bazze** [comment]
> @Bazze Did you find any workaround on this issue?

I did my own workaround, as I mentioned here: https://github.com/bpampuch/pdfmake/issues/441#issuecomment-161433196

Basically what I did back then was to rewrite the document measurement function to split the table when it got wider than the page width. However, it's not a solution that I've maintained or even using anymore. We've now redesigned the report we used it for so we no longer get very long tables.

**Comment by MussaCharles** [comment]
> > @Bazze Did you find any workaround on this issue?
> 
> I did my own workaround, as I mentioned here: [#441 (comment)](https://github.com/bpampuch/pdfmake/issues/441#issuecomment-161433196)
> 
> Basically what I did back then was to rewrite the document measurement function to split the table when it got wider than the page width. However, it's not a solution that I've maintained or even using anymore. We've now redesigned the report we used it for so we no longer get very long tables.

Thanks,  I  think we  might have to do a bit of design changes on our side too to avoid this issue.

**Comment by temo-o** [comment]
You can find answer in [this thread](https://github.com/bpampuch/pdfmake/issues/530).
It's basically this: 
    `pageSize: 'A0',//A0 is the largest A5 smallest(A0,A1,A2,A3,legal,A4,A5,letter))`

And if after this columns/data will be missaligned you can use this: ([src](https://stackoverflow.com/questions/33998231/how-to-change-export-pdf-font-size-in-datatables))

    `doc.defaultStyle.alignment = 'center';
     doc.styles.tableHeader.alignment = 'center';`

**Comment by danielpowell4** [comment]
It's a bit hacky of a solution but I found that if you first call `pdfMake.createPdf(...)` on the content you want, the library injects some real calculations into 'widths' rather than just relying on '*' and 'auto' 

From here we then loop and downsize the font to fit the content the the page per something like

```ts
// #region helpers
const MIN_FONT_SIZE = 4; // really small!
const buildPaddingLeft = (padding = 8) => (index: number) => index === 0 ? 0 : padding;
const buildPaddingRight = (padding = 8) => (index: number, node: ContentTable) => index + 1 === node.table.widths?.length ? 0 : padding;
const inchToPts = (inch: number) => inch * 72; // 1 inch = 72 points
// #endregion

// #region base case setup
const orientation = foo ? "portrait" : "landscape";
const pageWidth = orientation === "landscape" ? inchToPts(11) : inchToPts(8.5); // based on US Letter (8.5" x 11")
const pageSideMargin = orientation ? 40 : 80; // 20pts | 40pts each side
const maxPageWidth = pageWidth - pageSideMargin;
const tableLayout: TableLayout = { ...pdfMake.tableLayouts.default };
const initialFontSize = orientation ? 20 : 10; // or something different for you!
let lineItemsFontSize = initialFontSize;
const fontSizeDelta = Math.max(Math.round(initialFontSize / 10), 1); // the downsize step per run
// #endregion

// #region loop to fit table
let doesTableFitOnPage = false;
while (!doesTableFitOnPage) {
  doesTableFitOnPage = await new Promise<boolean>((resolve) =>
    pdfMake
      .createPdf({
        content: {
          ...lineItemContent,
          fontSize: lineItemsFontSize,
          layout: tableLayout,
        },
      })
      .getDataUrl(() => {
        if (lineItemsFontSize <= MIN_FONT_SIZE) return resolve(true); // too small... overflow

        const paddingRight = lineItemsFontSize - 2;
        const paddingLeft = lineItemsFontSize - 2;
        const paddingPerColumn = paddingRight + paddingLeft + 1;
        const edgeSkips = paddingLeft + paddingRight; // left and right most cells skip padding
        const paddingTotal =
          paddingPerColumn * lineItemContent.table.widths.length - edgeSkips;
        // @ts-expect-error _calcWidth private + injected by pdfmake.createPdf
        const colWidths = lineItemContent.table.widths.map((w) => w._calcWidth);
        const colTotal = colWidths.reduce((total, w) => (total += w), 0);
        const tableWidth = paddingTotal + colTotal;

        const tableFits = tableWidth <= maxPageWidth;
        if (tableFits) return resolve(true);

        // table does not fit, reduce font size and try again
        lineItemsFontSize -= fontSizeDelta;
        tableLayout.paddingLeft = buildPaddingLeft(
          lineItemsFontSize - fontSizeDelta,
        );
        tableLayout.paddingRight = buildPaddingRight(
          lineItemsFontSize - fontSizeDelta,
        );
        return resolve(false);
      }),
  );
}
// #endregion

// return content to where injected into document def
return { ...lineItemContent, fontSize: lineItemsFontSize, layout: tableLayout }; 
```
